File: backend/rules_engine/win_conditions.py
"""Win condition inspection. Win/elimination/back-row checks are applied as
part of moves.apply_move(); this module exposes a read-only view of that
result plus a standalone helper for checking the no-legal-moves condition."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from .game_state import GameState
from .moves import generate_legal_moves


@dataclass(frozen=True)
class WinResult:
    winner: str
    reason: str  # "back_row" | "elimination" | "no_legal_moves"


def check_winner(state: GameState) -> Optional[WinResult]:
    if state.winner is None:
        return None
    return WinResult(winner=state.winner, reason=state.win_reason or "")


def has_no_legal_moves(state: GameState, player: str | None = None) -> bool:
    player = player or state.current_player
    return len(generate_legal_moves(state, player)) == 0

File: frontend/src/lib/surge/engine/winConditions.ts
// Ported 1:1 from backend/rules_engine/win_conditions.py.
// Win/elimination/back-row checks are applied as part of moves.applyMove();
// this module exposes a read-only view of that result plus a standalone
// helper for checking the no-legal-moves condition.

import type { Player } from "./board";
import type { GameState } from "./gameState";
import { generateLegalMoves } from "./moves";

export interface WinResult {
  winner: Player;
  reason: string; // "back_row" | "elimination" | "no_legal_moves"
}

export function checkWinner(state: GameState): WinResult | null {
  if (state.winner === null) return null;
  return { winner: state.winner, reason: state.winReason ?? "" };
}

export function hasNoLegalMoves(state: GameState, player?: Player): boolean {
  const p = player ?? state.currentPlayer;
  return generateLegalMoves(state, p).length === 0;
}
